er_wheels/src/er_wheel_odometry_node.cpp

```diff
@@ -0,0 +1,118 @@
+  // check e.g. http://www8.cs.umu.se/kurser/5DV122/HT13/material/Hellstrom-ForwardKinematics.pdf for the formulas
+  nav_msgs::Odometry odometry_msg;
+  const double wheel_circumference = 2*M_PI*wheel_radius_;
+
+  odometry_msg.header.frame_id = "/odom";
+  odometry_msg.child_frame_id = "/base_link";
+  odometry_msg.header.stamp = ros::Time::now();
+  odometry_msg.twist.twist.angular.z = (angular_velocity_[MotorSide::Right] - angular_velocity_[MotorSide::Left]) * wheel_radius_ / ( 2 * base_radius_ );
+  odometry_msg.twist.twist.linear.x = (angular_velocity_[MotorSide::Right] + angular_velocity_[MotorSide::Left]) * wheel_radius_ / 2;
+
+  // how much time has passed since the last last and current update
+  const double dt = (odometry_msg.header.stamp - last_odometry_msg_.header.stamp).toSec();
+
+  // since we dont know when exactly between the last and current the velocity has changed, we take the average of the two measurements
+  const double v_avg = (odometry_msg.twist.twist.linear.x + last_odometry_msg_.twist.twist.linear.x) / 2;
+  const double omega_avg = (odometry_msg.twist.twist.angular.z + last_odometry_msg_.twist.twist.angular.z) / 2;
+  const double delta_omega = omega_avg * dt;
+
+  // TODO: any nicer way of doing that, using the old message is super annoying because of quaternions...?
+  // make sure yaw_ is in [0,2pi)
+  yaw_ = yaw_ + delta_omega;
+  yaw_ = yaw_ >= 2 * M_PI ? yaw_ - 2 * M_PI : yaw_;
+  yaw_ = yaw_ < 0 ? yaw_ + 2 * M_PI : yaw_;
+  tf2::Quaternion new_rotation;
+  new_rotation.setRPY(0, 0, yaw_);
+  odometry_msg.pose.pose.orientation = tf2::toMsg(new_rotation);
+
+  // find radius and center of the circle trajectory that the robot has been following the past dt
+  // TODO: different formulas when omega_avg is very small -> infinite circle / driving in a straight line
+  if(omega_avg == 0)
+    return;
+  const double r = v_avg / omega_avg; // TODO: check for omega_avg == 0? required?
+  const double icc_x = last_odometry_msg_.pose.pose.position.x - r * sin(yaw_);
+  const double icc_y = last_odometry_msg_.pose.pose.position.y + r * cos(yaw_);
+  const double delta_x = last_odometry_msg_.pose.pose.position.x - icc_x;
+  const double delta_y = last_odometry_msg_.pose.pose.position.y - icc_y;
+
+  ROS_INFO("delta_omega: %f", delta_omega);
+  ROS_INFO("delta_x: %f", delta_x);
+  ROS_INFO("delta_y: %f", delta_y);
+  ROS_INFO("r %f", r);
+  ROS_INFO("yaw %f", yaw_);
+  ROS_INFO("icc_x %f", icc_x);
+  ROS_INFO("icc_y %f", icc_y);
+
+  odometry_msg.pose.pose.position.x = icc_x + cos(delta_omega) * delta_x - sin(delta_omega) * delta_y;
+  odometry_msg.pose.pose.position.y = icc_y + sin(delta_omega) * delta_x + cos(delta_omega) * delta_y;
+
+  last_odometry_msg_ = odometry_msg;
+  odometry_pub_.publish(odometry_msg);
+}
+
+int main(int argc, char **argv)
+{
+  ros::init(argc, argv, "er_wheel_odometry_node");
+
+  WheelOdometryNode wheel_odometry_node;
+
+  wheel_odometry_node.run_node();
```

Comment:
I don't understand why the whileloop is run inside a function. I see that you want to keep the main() clean of subs etc, but why not have them declared as private? the run_node() method looks a bit crazy to me. but hey, should work anyway ;)